### PR TITLE
add reduce=True argument to MultiLabelMarginLoss

### PR DESCRIPTION
**Pull request by li-roy**
As per #264. When reduce is False, MultiLabelMarginLoss outputs a loss per sample in minibatch. When reduce is True (default), the current behavior is kept.

Test Plan
test/run_test.sh
Added unit test. For the reduce=False case. Added unit tests for 1d tensors. Added a reference function.

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -565,10 +565,32 @@ class MultiLabelMarginLoss(_Loss):
...
+        reduce (bool, optional): By default, the losses are averaged or summed over
+            observations for each minibatch depending on size_average. When reduce
+            is False, returns a loss per batch element instead and ignores
+            size_average. Default: True
+
+    Shape:
+        - Input: :math:`(N)` or :math:`(N, *)` where `*` means, any number of additional
+          dimensions
+        - Target: :math:`(N)` or :math:`(N, *)`, same shape as the input
+        - Output: scalar. If `reduce` is False, then `(N)` or `(N, *)`, same shape as
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: If reduce is False, isn't the output always of size `(N)` ?

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -565,10 +565,32 @@ class MultiLabelMarginLoss(_Loss):
     The criterion only considers the first non-negative `y[j]` targets.
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: Did we ever figure out if this means anything? Should this be deleted?

File: test/test_nn.py [hunk 3]
```diff
@@ -4557,6 +4557,34 @@ def smoothl1loss_no_reduce_test():
...
+        input_fn=lambda: torch.randn(10),
+        reference_fn=lambda i, _:
+            loss_reference_fns['MultiLabelMarginLoss'](i, t.data.type_as(i).long(), reduce=False),
+        check_no_size_average=True,
+        check_gradgrad=False,
+        pickle=False)
+
+
+def multilabelmarginloss_no_reduce_test():
+    t = Variable(torch.rand(5, 10).mul(10).floor().long())
```
Comment (zou3519) [inline test/test_nn.py]: You should add another test where you test a negative index in the target

File: test/common_nn.py [hunk 4]
```diff
@@ -439,10 +483,22 @@ def smoothl1loss_reference(input, target, size_average=True, reduce=True):
         desc='margin',
         check_no_size_average=True,
     ),
+    dict(
+        module_name='MultiLabelMarginLoss',
```
Comment (zou3519) [inline test/common_nn.py]: Thanks for adding a test for the 1d case!

File: test/common_nn.py [hunk 5]
```diff
@@ -322,11 +322,55 @@ def smoothl1loss_reference(input, target, size_average=True, reduce=True):
...
+    return sum
+
+
+def multilabelmarginloss_reference(input, target, size_average=True, reduce=True):
+    is_target = torch.LongTensor(input.size()).zero_()
+
+    if input.dim() == 1:
+        n = 1
+        dim = input.size()[0]
+        output = torch.Tensor(n).zero_()
```
Comment (zou3519) [inline test/common_nn.py]: nit: this is better written as `torch.zeros(n)`. Same with the other two places in this function

File: test/common_nn.py [hunk 6]
```diff
@@ -322,11 +322,55 @@ def smoothl1loss_reference(input, target, size_average=True, reduce=True):
     return output
 
 
+def _multilabelmarginloss_reference(input, target, is_target):
```
Comment (zou3519) [inline test/common_nn.py]: nit: You can nest this function inside of `multilabelmarginloss_reference` so that it is only visible to the scope of that function but it doesn't really matter

File: test/common_nn.py [hunk 7]
```diff
@@ -322,11 +322,55 @@ def smoothl1loss_reference(input, target, size_average=True, reduce=True):
...
+
+    return sum
+
+
+def multilabelmarginloss_reference(input, target, size_average=True, reduce=True):
+    is_target = torch.LongTensor(input.size()).zero_()
+
+    if input.dim() == 1:
+        n = 1
+        dim = input.size()[0]
```
Comment (zou3519) [inline test/common_nn.py]: nit: I think you can do `input.size(0)`

File: test/common_nn.py [hunk 8]
```diff
@@ -322,11 +322,55 @@ def smoothl1loss_reference(input, target, size_average=True, reduce=True):
     return output
 
 
+def _multilabelmarginloss_reference(input, target, is_target):
+    sum = 0
+    for i in range(0, target.size()[0]):
+        target_index = target[i]
```
Comment (zou3519) [inline test/common_nn.py]: It's a little hard to understand this (I understand that the THNN implementation uses something similar). Is it possible to rewrite this to look more like the formula? Maybe capture the `i != y[j]` part without using an `is_target` array? It's okay if it's not as efficient since we want this function to only be used in testing.

**Review by zou3519 — COMMENTED** [review]
Haven't reviewed the C code or kernels yet, will get to that soon. I left a few comments on the python side to start out

File: aten/src/THNN/generic/MultiLabelMarginCriterion.c [hunk 9]
```diff
@@ -176,10 +226,33 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
     isTarget_data += dim;
     gradInput_data += dim;
   }
+  gradInput_data -= nframe*dim;
```
Comment (zou3519) [inline aten/src/THNN/generic/MultiLabelMarginCriterion.c]: You should do `gradInput_data = THTensor_(data)(gradInput)` instead so the code is more readable.

Could you assert that THTensor_(isContiguous)(gradInput) is true? The following code works under this assumption and I'm not 100% sure gradInput will always be contiguous, but I'll get back to you on that.

File: aten/src/THNN/generic/MultiLabelMarginCriterion.c [hunk 10]
```diff
@@ -76,17 +126,15 @@ void THNN_(MultiLabelMarginCriterion_updateOutput)(
         }
       }
     }
+
+    sum /= dim;
+    THTensor_(set1d)(output, t, sum);
```
Comment (zou3519) [inline aten/src/THNN/generic/MultiLabelMarginCriterion.c]: There's a `THTensor_fastSet1d` I think

File: aten/src/THNN/generic/MultiLabelMarginCriterion.c [hunk 11]
```diff
@@ -176,10 +226,33 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
...
+    THNN_CHECK_DIM_SIZE(gradOutput, 1, 0, 1);
+    for (t = 0; t < nframe*dim; t++)
+    {
+      gradInput_data[t] *= THTensor_fastGet1d(gradOutput, 0);
+    }
+  }
+  else
+  {
+    THNN_CHECK_DIM_SIZE(gradOutput, 1, 0, nframe);
+    gradOutput = THTensor_(newContiguous)(gradOutput);
```
Comment (zou3519) [inline aten/src/THNN/generic/MultiLabelMarginCriterion.c]: Is this necessary? The `fastGet1d` method should take into account the strides of `gradOutput`, right?

File: aten/src/THNN/generic/MultiLabelMarginCriterion.c [hunk 12]
```diff
@@ -176,10 +226,33 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
...
+      {
+        gradInput_data[t * dim + d] *= THTensor_fastGet1d(gradOutput, t);
+      }
+    }
+  }
 
   THTensor_(free)(input);
   THIndexTensor_(free)(target);
   THTensor_(free)(isTarget);
+  THTensor_(free)(gradOutput);
```
Comment (zou3519) [inline aten/src/THNN/generic/MultiLabelMarginCriterion.c]: You should only free gradOutput inside the `else` statement because that's where you made a copy of it

**Review by zou3519 — COMMENTED** [review]
The THNN code looks good for the most part! Just a few nits here and there. Haven't looked at the cuda kernel yet.

File: aten/src/THCUNN/MultiLabelMarginCriterion.cu [hunk 13]
```diff
@@ -77,12 +77,14 @@ __global__ void cunn_MultiLabelMarginCriterion_updateOutput_kernel(Dtype *output
 
 template <typename Dtype, typename Acctype>
 __global__ void cunn_MultiLabelMarginCriterion_updateGradInput_kernel(Dtype *gradInput,
+                                                                      Dtype *gradOutput,
```
[the remark this answers is not in the document]
Comment (li-roy) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: This isn't really consistent in the first place, for example gradInput is the first arg here. If it matters, I can completely reorder these args.

File: torch/nn/modules/loss.py [hunk 14]
```diff
@@ -562,13 +562,35 @@ class MultiLabelMarginLoss(_Loss):
...
+
+    Args:
+        size_average (bool, optional): By default, the losses are averaged
+            over observations for each minibatch. However, if the field
+            size_average is set to ``False``, the losses are instead summed for
+            each minibatch. Default: ``True``
+        reduce (bool, optional): By default, the losses are averaged or summed over
+            observations for each minibatch depending on size_average. When reduce
+            is False, returns a loss per batch element instead and ignores
+            size_average. Default: True
```
Comment (SsnL) [inline torch/nn/modules/loss.py]: nit: \`\`True\`\`

File: aten/src/THCUNN/MultiLabelMarginCriterion.cu [hunk 15]
```diff
@@ -133,6 +140,11 @@ __global__ void cunn_MultiLabelMarginCriterion_updateGradInput_kernel(Dtype *gra
     }
     __syncthreads();
```
Comment (SsnL) [inline aten/src/THCUNN/MultiLabelMarginCriterion.cu]: Are this syncthreads and the one three lines below necessary?

File: torch/nn/modules/loss.py (hunk 14 above)
Comment (SsnL) [inline torch/nn/modules/loss.py]: nit: \`\`True\`\`

File: torch/nn/modules/loss.py [hunk 16]
```diff
@@ -562,13 +562,35 @@ class MultiLabelMarginLoss(_Loss):
...
 
     This allows for different samples to have variable amounts of target classes
+
+    Args:
+        size_average (bool, optional): By default, the losses are averaged
+            over observations for each minibatch. However, if the field
+            size_average is set to ``False``, the losses are instead summed for
+            each minibatch. Default: ``True``
+        reduce (bool, optional): By default, the losses are averaged or summed over
+            observations for each minibatch depending on size_average. When reduce
```
Comment (SsnL) [inline torch/nn/modules/loss.py]: nit: It'd be great if you can write all argument in this style: :attr:\`size_average\`

File: aten/src/THCUNN/MultiLabelMarginCriterion.cu (hunk 15 above)
Comment (li-roy) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: I think you're right about the one below. I'll take that out. Correct me if I'm wrong, but isn't this one needed so that sum doesn't change on other threads when the reduce is being run?

File: aten/src/THNN/generic/MultiLabelMarginCriterion.c [hunk 17]
```diff
@@ -252,7 +252,7 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
   THTensor_(free)(input);
   THIndexTensor_(free)(target);
   THTensor_(free)(isTarget);
-  THTensor_(free)(gradOutput);
+  THTensor_(free)(gradInput);
```
Comment (zou3519) [inline aten/src/THNN/generic/MultiLabelMarginCriterion.c]: You shouldn't free gradInput -- I don't think you've made a copy of it.

File: aten/src/THCUNN/MultiLabelMarginCriterion.cu (hunk 15 above)
Comment (zou3519) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: On second thought, I think it's necessary so that `gradInput_k[target_idx]` isn't being written to by multiple threads.

Comment (SsnL) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: @li-roy I think the syncthreads before the reduction makes sure that data is available at all threads when doing reduction.

@zou3519 I think only thread 0 writes to `gradInput_k[target_idx]`

Comment (li-roy) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: @SsnL But the syncthreads after the reduction is needed so that each thread doesn't continue to the next iteration of the loop and overwrite sum before the reduction is completed on thread 0.

File: aten/src/THNN/generic/MultiLabelMarginCriterion.c (hunk 17 above)
Comment (li-roy) [reply aten/src/THNN/generic/MultiLabelMarginCriterion.c]: Haven't I now made a copy at line 196?

Comment (zou3519) [reply aten/src/THNN/generic/MultiLabelMarginCriterion.c]: Oops sorry I misspelled that!

File: aten/src/THCUNN/MultiLabelMarginCriterion.cu (hunk 15 above)
Comment (SsnL) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: Yes, `sum` may be overwritten. But in `reduceBlock`, the values are first into shared memory (`sums` here) and a `__syncthreads()` is called (see https://github.com/pytorch/pytorch/blob/502aaf39cf4a878f9e4f849e5f409573aa598aa9/aten/src/THC/THCReduceApplyUtils.cuh#L39-L48). Since there is `__syncthreads()` before the reduction, the `sums` shared memory won't be overwritten if I'm understanding correctly.

Comment (li-roy) [reply aten/src/THCUNN/MultiLabelMarginCriterion.cu]: Oh okay, in that case I agree. Thanks for the pointer!

**Review by zou3519 — APPROVED** [review]
All comments addressed, LGTM unless anyone has something else to add. Thanks @li-roy!

**Comment by SsnL** [comment]
Thanks @li-roy !

**Comment by zou3519** [comment]
I'm not sure why the CI isn't running, but... let's give this one final test
